### ISSUE TITLE
Error: Cannot instantiate underlying libxml2 structure

### ISSUE THREAD

**Issue by grwlf** [loskutov/deadbeef-lyricbar#10]
Hi. Thanks for nice plugin! I've managed to compile and run it, btw default.nix really helps. Unfortunately, plugin can't find any lyrics. Every song  make it issue something like the following

```
lyricbar: couldn't parse XML (URI is 'http://lyrics.wikia.com/api.php?action=lyrics&fmt=xml&artist=Tina%20Turner&song=Golden%20Eye'), what(): Cannot instantiate underlying libxml2 structure
```

**Comment by grwlf** [comment]
Versions on my system

```
[nix-shell:~/proj/deadbeef-lyricbar]$ ldd ddb_lyricbar_gtk3.so 
	linux-vdso.so.1 (0x00007fe7e85c6000)
	libxml++-3.0.so.1 => /nix/store/sh2j1jh62mxi3nqcbp4j587whk5lgi2a-libxml++-3.0.1/lib/libxml++-3.0.so.1 (0x00007fe7e8561000)
	libxml2.so.2 => /nix/store/fwjm8qsahpgd9qygvaiwsjwyap4nrf2k-libxml2-2.9.9/lib/libxml2.so.2 (0x00007fe7e83f7000)
	libgtkmm-3.0.so.1 => /nix/store/96klax2i1imr0y246wn773c25c44bfig-gtkmm-3.24.1/lib/libgtkmm-3.0.so.1 (0x00007fe7e7e3a000)
	libatkmm-1.6.so.1 => /nix/store/d0i448p3dv544wn97h8kyx7vn6vppd2k-atkmm-2.28.0/lib/libatkmm-1.6.so.1 (0x00007fe7e7de6000)
	libgdkmm-3.0.so.1 => /nix/store/96klax2i1imr0y246wn773c25c44bfig-gtkmm-3.24.1/lib/libgdkmm-3.0.so.1 (0x00007fe7e7d86000)
	libgiomm-2.4.so.1 => /nix/store/lmqx0ixc53q0piyx00khr5jg7h7x2zad-glibmm-2.60.0/lib/libgiomm-2.4.so.1 (0x00007fe7e7bad000)
	libpangomm-1.4.so.1 => /nix/store/k3dll6yy6v0pav734q1mnrf349pa9ihi-pangomm-2.42.0/lib/libpangomm-1.4.so.1 (0x00007fe7e7b78000)
	libglibmm-2.4.so.1 => /nix/store/lmqx0ixc53q0piyx00khr5jg7h7x2zad-glibmm-2.60.0/lib/libglibmm-2.4.so.1 (0x00007fe7e7ae4000)
	libcairomm-1.0.so.1 => /nix/store/k5h9vq6x3mm9hag3ws3c7jy4dwzr55jb-cairomm-1.12.2/lib/libcairomm-1.0.so.1 (0x00007fe7e7ab8000)
	libsigc-2.0.so.0 => /nix/store/fkmqwgly2n4kfm7hn5nidj6nq8bzr35f-libsigc++-2.10.1/lib/libsigc-2.0.so.0 (0x00007fe7e7aaf000)
	libgtk-3.so.0 => /nix/store/m7yzspvdc1a1hvwy36ibpll5ckmg9kn0-gtk+3-3.24.10/lib/libgtk-3.so.0 (0x00007fe7e7398000)
	libgdk-3.so.0 => /nix/store/m7yzspvdc1a1hvwy36ibpll5ckmg9kn0-gtk+3-3.24.10/lib/libgdk-3.so.0 (0x00007fe7e72a4000)
	libpangocairo-1.0.so.0 => /nix/store/6qahh6icgl2v6wyyfgr1ar8p5fbdgkdj-pango-1.43.0/lib/libpangocairo-1.0.so.0 (0x00007fe7e7294000)
	libpango-1.0.so.0 => /nix/store/6qahh6icgl2v6wyyfgr1ar8p5fbdgkdj-pango-1.43.0/lib/libpango-1.0.so.0 (0x00007fe7e7248000)
	libatk-1.0.so.0 => /nix/store/h4r6dzai15zchhmry3g51bw52vqir3p3-atk-2.32.0/lib/libatk-1.0.so.0 (0x00007fe7e721f000)
	libcairo-gobject.so.2 => /nix/store/36rnjvyjyf9nn2nxz71w1xrvqgj70a8q-cairo-1.16.0/lib/libcairo-gobject.so.2 (0x00007fe7e7213000)
	libcairo.so.2 => /nix/store/36rnjvyjyf9nn2nxz71w1xrvqgj70a8q-cairo-1.16.0/lib/libcairo.so.2 (0x00007fe7e70db000)
	libgdk_pixbuf-2.0.so.0 => /nix/store/rfn9sknqqjllxs75xckzxrkhfvl1n31p-gdk-pixbuf-2.38.1/lib/libgdk_pixbuf-2.0.so.0 (0x00007fe7e70b4000)
	libgio-2.0.so.0 => /nix/store/sn9f38pp0qkqqbbarc7jbrqi3mgwxim0-glib-2.60.7/lib/libgio-2.0.so.0 (0x00007fe7e6ef4000)
	libgobject-2.0.so.0 => /nix/store/sn9f38pp0qkqqbbarc7jbrqi3mgwxim0-glib-2.60.7/lib/libgobject-2.0.so.0 (0x00007fe7e6e96000)
	libglib-2.0.so.0 => /nix/store/sn9f38pp0qkqqbbarc7jbrqi3mgwxim0-glib-2.60.7/lib/libglib-2.0.so.0 (0x00007fe7e6d73000)
	libstdc++.so.6 => /nix/store/bi13kcq64lf1b88sm7k3siskyi63qhak-gcc-8.3.0-lib/lib/libstdc++.so.6 (0x00007fe7e6be8000)
	libm.so.6 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libm.so.6 (0x00007fe7e6a52000)
	libgcc_s.so.1 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libgcc_s.so.1 (0x00007fe7e683c000)
	libc.so.6 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libc.so.6 (0x00007fe7e6686000)
	libdl.so.2 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libdl.so.2 (0x00007fe7e667f000)
	libz.so.1 => /nix/store/i6y4i4s55ksqki62sx20jp6bk81crkic-zlib-1.2.11/lib/libz.so.1 (0x00007fe7e6660000)
	libgmodule-2.0.so.0 => /nix/store/sn9f38pp0qkqqbbarc7jbrqi3mgwxim0-glib-2.60.7/lib/libgmodule-2.0.so.0 (0x00007fe7e6659000)
	libfreetype.so.6 => /nix/store/nghs0bz0pwxlbd5r4xc9lkaxs1kvz24b-freetype-2.10.1/lib/libfreetype.so.6 (0x00007fe7e6597000)
	libpng16.so.16 => /nix/store/wb8k0b9aww55311669hzmcyhp3amhihr-libpng-apng-1.6.37/lib/libpng16.so.16 (0x00007fe7e655e000)
	libXext.so.6 => /nix/store/qyy2i7dslkdllwvmkkvnybfvb8d150l8-libXext-1.3.4/lib/libXext.so.6 (0x00007fe7e6549000)
	libXrender.so.1 => /nix/store/vjaxnzj2lqibqhn2l0aq133dp68f2701-libXrender-0.9.10/lib/libXrender.so.1 (0x00007fe7e653c000)
	libX11.so.6 => /nix/store/b654j443kmhgdgaxw3jyp67w1pyrd5xs-libX11-1.6.8/lib/libX11.so.6 (0x00007fe7e63f9000)
	libpangoft2-1.0.so.0 => /nix/store/6qahh6icgl2v6wyyfgr1ar8p5fbdgkdj-pango-1.43.0/lib/libpangoft2-1.0.so.0 (0x00007fe7e63dd000)
	libfontconfig.so.1 => /nix/store/hm1g7yzm8l75523cmfs3w9xdvnrd56ax-fontconfig-2.12.6-lib/lib/libfontconfig.so.1 (0x00007fe7e6395000)
	libfribidi.so.0 => /nix/store/pl4wl62hm4053vy3yk7j8bg2q2m53rw2-fribidi-1.0.5/lib/libfribidi.so.0 (0x00007fe7e6377000)
	libepoxy.so.0 => /nix/store/zhn886wryhs8kbb4rknn29vsnhh6ncv8-epoxy-1.5.3/lib/libepoxy.so.0 (0x00007fe7e623d000)
	libXi.so.6 => /nix/store/hvbwynhxnfc42ygjkdc4d2nihnwr8jpc-libXi-1.7.10/lib/libXi.so.6 (0x00007fe7e622a000)
	libatk-bridge-2.0.so.0 => /nix/store/0da2hizg4a0v79f0afx4y5kdwlp3mn9x-at-spi2-atk-2.32.0/lib/libatk-bridge-2.0.so.0 (0x00007fe7e61f4000)
	libXfixes.so.3 => /nix/store/yrfswgx448ilr920794v648ff6sardkx-libXfixes-5.0.3/lib/libXfixes.so.3 (0x00007fe7e61ec000)
	libpthread.so.0 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libpthread.so.0 (0x00007fe7e61cb000)
	librt.so.1 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/librt.so.1 (0x00007fe7e61bf000)
	libxkbcommon.so.0 => /nix/store/n37v8l2lfg129r8rdrz81pkn05464zj2-libxkbcommon-0.8.4/lib/libxkbcommon.so.0 (0x00007fe7e617d000)
	libwayland-client.so.0 => /nix/store/a7hnxwxdp33nbylyn59m3da85aj5irqh-wayland-1.17.0/lib/libwayland-client.so.0 (0x00007fe7e616b000)
	libwayland-cursor.so.0 => /nix/store/a7hnxwxdp33nbylyn59m3da85aj5irqh-wayland-1.17.0/lib/libwayland-cursor.so.0 (0x00007fe7e615f000)
	libwayland-egl.so.1 => /nix/store/a7hnxwxdp33nbylyn59m3da85aj5irqh-wayland-1.17.0/lib/libwayland-egl.so.1 (0x00007fe7e615a000)
	libXcursor.so.1 => /nix/store/4pghpydvdfba0ak4z0p3lzr9rbrcwlka-libXcursor-1.2.0/lib/libXcursor.so.1 (0x00007fe7e614e000)
	libXdamage.so.1 => /nix/store/4sna8632gqp2sqlr9n30wvnpshicm5vb-libXdamage-1.1.5/lib/libXdamage.so.1 (0x00007fe7e6149000)
	libXcomposite.so.1 => /nix/store/adil832h835py7sx1g3xx5yi08swq4py-libXcomposite-0.4.5/lib/libXcomposite.so.1 (0x00007fe7e6144000)
	libXrandr.so.2 => /nix/store/3pgidjsyngcg469cj2v6bnr4swv4dba6-libXrandr-1.5.2/lib/libXrandr.so.2 (0x00007fe7e6137000)
	libXinerama.so.1 => /nix/store/kq3qd9magcym7qwkd2shk354l1v3n4ih-libXinerama-1.1.4/lib/libXinerama.so.1 (0x00007fe7e6130000)
	libpixman-1.so.0 => /nix/store/jql6xxj0lcrgyzb3bnzbp6kzp9qka7mj-pixman-0.38.4/lib/libpixman-1.so.0 (0x00007fe7e6085000)
	libEGL.so.1 => /nix/store/5mv36m6ivbd3pw02i26dl5w39siwqyjh-libglvnd-1.0.0/lib/libEGL.so.1 (0x00007fe7e6070000)
	libxcb-shm.so.0 => /nix/store/2rwmc77f9kvfvmpg4z4r4b1cv3lxfsy6-libxcb-1.13.1/lib/libxcb-shm.so.0 (0x00007fe7e606b000)
	libxcb.so.1 => /nix/store/2rwmc77f9kvfvmpg4z4r4b1cv3lxfsy6-libxcb-1.13.1/lib/libxcb.so.1 (0x00007fe7e603e000)
	libxcb-render.so.0 => /nix/store/2rwmc77f9kvfvmpg4z4r4b1cv3lxfsy6-libxcb-1.13.1/lib/libxcb-render.so.0 (0x00007fe7e602e000)
	libGL.so.1 => /nix/store/5mv36m6ivbd3pw02i26dl5w39siwqyjh-libglvnd-1.0.0/lib/libGL.so.1 (0x00007fe7e5f9a000)
	libmount.so.1 => /nix/store/lk5w3vznbq8wxsmzv19bab1n6zw78lkh-util-linux-2.33.2/lib/libmount.so.1 (0x00007fe7e5f3a000)
	libselinux.so.1 => /nix/store/kk4m7hgr3b8wbhlcclkhbwwvylrba9yx-libselinux-2.9/lib/libselinux.so.1 (0x00007fe7e5f0b000)
	libresolv.so.2 => /nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib/libresolv.so.2 (0x00007fe7e5ef2000)
	libffi.so.6 => /nix/store/jagnydxh3hrrydfzq8q2khwjicskci42-libffi-3.2.1/lib/../lib64/libffi.so.6 (0x00007fe7e5ee6000)
	libpcre.so.1 => /nix/store/gp453gh080k5syqsa3cck0mpyrd89nib-pcre-8.43/lib/libpcre.so.1 (0x00007fe7e5e73000)
	/nix/store/w7hmngvdx4fhg3vd8iad0hjh58w3gqsm-glibc-2.27/lib64/ld-linux-x86-64.so.2 (0x00007fe7e85c8000)
	libbz2.so.1 => /nix/store/1krcqa0baqvvr41h4p5mhahm5cj5sv9n-bzip2-1.0.6.0.1/lib/libbz2.so.1 (0x00007fe7e5e5e000)
	libharfbuzz.so.0 => /nix/store/7l5rzz69b29v9wygdabwpqkcsb67i6w5-harfbuzz-2.6.1/lib/libharfbuzz.so.0 (0x00007fe7e5d61000)
	libexpat.so.1 => /nix/store/vzw3injdwqd0j87n7pmnas1q2ksiccwc-expat-2.2.7/lib/libexpat.so.1 (0x00007fe7e5d24000)
	libdbus-1.so.3 => /nix/store/azxgwa0xbzmvpw6may2fqf7ywa9la0jx-dbus-1.12.16-lib/lib/libdbus-1.so.3 (0x00007fe7e5cce000)
	libatspi.so.0 => /nix/store/gagx45fcjldrd1ilc9c8sl540800sjh2-at-spi2-core-2.32.1/lib/libatspi.so.0 (0x00007fe7e5c98000)
	libGLdispatch.so.0 => /nix/store/5mv36m6ivbd3pw02i26dl5w39siwqyjh-libglvnd-1.0.0/lib/libGLdispatch.so.0 (0x00007fe7e5bdf000)
	libXau.so.6 => /nix/store/j904j2yzd4ym620xk7cqc3aymgh9x07n-libXau-1.0.9/lib/libXau.so.6 (0x00007fe7e5bda000)
	libXdmcp.so.6 => /nix/store/79f00fz2pkk3blagyb3q11yasrpnlycy-libXdmcp-1.1.3/lib/libXdmcp.so.6 (0x00007fe7e5bd0000)
	libGLX.so.0 => /nix/store/5mv36m6ivbd3pw02i26dl5w39siwqyjh-libglvnd-1.0.0/lib/libGLX.so.0 (0x00007fe7e5b9b000)
	libblkid.so.1 => /nix/store/lk5w3vznbq8wxsmzv19bab1n6zw78lkh-util-linux-2.33.2/lib/libblkid.so.1 (0x00007fe7e5b45000)
	libuuid.so.1 => /nix/store/lk5w3vznbq8wxsmzv19bab1n6zw78lkh-util-linux-2.33.2/lib/libuuid.so.1 (0x00007fe7e5b3c000)
	libgraphite2.so.3 => /nix/store/x4qsd2qgg4i8m3kvvin7hg98wmqxkk0g-graphite2-1.3.6/lib/libgraphite2.so.3 (0x00007fe7e5b09000)
	libsystemd.so.0 => /nix/store/sswf2av8cgvna98qlfdp248z4888d0fs-systemd-243-lib/lib/libsystemd.so.0 (0x00007fe7e5a4f000)
	liblzma.so.5 => /nix/store/npv7xjzp0bpif1xpbzhj08h1ygpx07rs-xz-5.2.4/lib/liblzma.so.5 (0x00007fe7e5a25000)
	liblz4.so.1 => /nix/store/28fqr80w8hphpm2hd4qj3ifgh6dnynsr-lz4-1.9.1/lib/liblz4.so.1 (0x00007fe7e59ef000)
	libcap.so.2 => /nix/store/5vw9abs50y4i7sx2ymcmh2sm7y691y4x-libcap-2.27-lib/lib/libcap.so.2 (0x00007fe7e59e8000)
	libgcrypt.so.20 => /nix/store/8qrq8ajydrz17l5aywzs5hxbc2sbp9qa-libgcrypt-1.8.5/lib/libgcrypt.so.20 (0x00007fe7e58c9000)
	libgpg-error.so.0 => /nix/store/vvar6smw7zcx8fshv8qkgh4pw38m3w5c-libgpg-error-1.36/lib/libgpg-error.so.0 (0x00007fe7e58a5000)
```

**Comment by loskutov** [comment]
I can confirm I observe the same on my machine. It looks like Wikia started to force HTTPS for their API, which libxml2 doesn't have built-in support for. Hopefully I'll find time to come up with a fix...

**Comment by dakeryas** [comment]
Thanks for looking into the fix! 

I was tracking down the bug to the failed initialisation of the `xmlNewTextReaderFilename _imlf` member in  `libxml++-3.0.1/libxml++/parsers/textreader.cc` without proceeding further seeing that this was delegated to libxml2 indeed... 

I assume there is no easy port to another more modern XML parser, is there?

**Comment by loskutov** [comment]
Minimizing number of dependencies was an intentional design decision (worth noting that libxml++3 is still for some reason not in Debian repos, which is kind of surprising for me: it's been almost 3 years since 3.0.1 release, and they're still shipping 2.40.1 only). I think the fix is to use Glib for downloading the page and then passing its contents to libxml++, instead of just passing the URL.

**Comment by dakeryas** [comment]
Yes, I was surprised to have to build libxml3 myself as well...

Ah, I think I understand, so you would just try and use the `TextReader` constructor which takes the data pointer `const unsigned char* data`, which you would have filled yourself from the URL!

**Comment by loskutov** [comment]
Fixed in 003b85a508dff80910978090b1e67f55bc2ef437.

**Comment by dakeryas** [comment]
Thanks a lot, that is working for me!

Just a proposed warning fix here:

```diff --git a/src/utils.cpp b/src/utils.cpp
index d51c307..201cdde 100644
--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -214,7 +214,7 @@ experimental::optional<ustring> download_lyrics_from_lyricwiki(DB_playItem_t *tr
                return {};
        }
        try {
-               xmlpp::TextReader reader{(const unsigned char *)doc->data(), doc->size()};
+               xmlpp::TextReader reader{(const unsigned char *)doc->data(), static_cast<unsigned int>(doc->size())};
 
                while (reader.read()) {
                        if (reader.get_node_type() == xmlpp::TextReader::NodeType::Element
@@ -246,7 +246,7 @@ experimental::optional<ustring> download_lyrics_from_lyricwiki(DB_playItem_t *tr
        }
        string raw_lyrics;
        try {
-               xmlpp::TextReader reader{(const unsigned char *)doc->data(), doc->size()};
+               xmlpp::TextReader reader{(const unsigned char *)doc->data(), static_cast<unsigned int>(doc->size())};
                while (reader.read()) {
                        if (reader.get_name() == "rev") {
                                reader.read();
```

I tried to push to a new branch to issue a merge request, but it would not let me authenticate :see_no_evil:
